ckanext/unaids/tests/test_giftless_backend.py

```diff
@@ -48,3 +53,23 @@ def test_giftless_resource_create(self):
         assert 'sha256' in resource
         assert 'lfs_prefix' in resource
         assert resource.get('name', None) == filename
+
+
+    def test_resource_url_encoding_test(self):
+        user = factories.User()
+        org = factories.Organization(
+            users=[
+                {'name': user['name'], 'capacity': 'admin'},
+            ]
+        )
+        dataset = factories.Dataset(user=user, owner_org=org['id'])
+        filename = 'file%.csv'
+        resource = helpers.call_action('resource_create', package_id=dataset["id"], name=filename)
+        url = h.url_for('resource.view',
+                                    id=dataset['id'],
+                                    resource_id=resource['id'],
+                                    filename=filename,
+                                    qualified=True)
+
+        url_not_encoded = unquote(url)
+        assert unaids_helpers.url_encode(url_not_encoded) == url
```

Comment:
```
      url_not_encoded = unquote(url)
      assert unaids_helpers.url_encode(url_not_encoded) == url
```
This assertion is practically testing `urllib.parse` library instead of our code. First, you unquoting (with `urllib.prase.unquote`) and then testing if `url_encode` (which uses `urllib.parse.quote`) quotes it properly.

Please have a look at this test as an example on how you can test your bugfix:
`ckanext.unaids.tests.test_dataset_releases.TestDatasetRead._get_dataset_release_sidebar`